src/js/common/validator.js

```diff
@@ -1,122 +1,540 @@
+     * @return {String} - A message key for the given error, or the unaltered default message if no message key is found.
+     *
+     */
+    gpii.schema.validator.errorHintForRule = function (rulePath, gssSchema, defaultMessage) {
+        defaultMessage = defaultMessage || gpii.schema.validator.defaultI18nKeysByRule["schema-validator-general-failure"];
+        var enclosingDefinitionSegments = rulePath.slice(0, -1);
+        var finalRuleSegment = rulePath[rulePath.length - 1];
+
+        var errorsDef = fluid.get(gssSchema, enclosingDefinitionSegments.concat("errors"));
+        // "short" notation.
+        if (typeof errorsDef === "string") {
+            return errorsDef;
+        }
+        // "long" notation.
+        else {
+            var ruleErrorKey = fluid.get(errorsDef, finalRuleSegment);
+            if (ruleErrorKey) {
+                return ruleErrorKey;
+            }
+            else {
+                var fieldErrorKey = fluid.get(errorsDef, [""]);
+                if (fieldErrorKey) { return fieldErrorKey; }
+            }
+        }
+
+        // If we haven't found any information to help provide an error message key, look for a global default for the
+        // rule type.  If none is found, return the underlying message from AJV.
+        return gpii.schema.validator.defaultI18nKeysByRule[finalRuleSegment] || defaultMessage;
+    };
+
+    /**
+     *
+     * Standardise any AJV errors received so that they return our standard format.  If there are no validation errors,
+     * the return value should look like:
+     *
+     * { isValid: true }
+     *
+     * If there are validation errors, the return value should look like:
+     *
+     * {
+     *  isValid: false,
+     *  errors: [
+     *      {
+     *          dataPath: ["requiredField"],
+     *          schemaPath: ["properties", "requiredField", "required"],
+     *          rule: { required: true },
+     *          message: "validation-required-field-missing"
+     *      },
+     *      {
+     *          dataPath: ["deep", "booleanField"],
+     *          schemaPath: ["properties", "deep", "properties", "booleanField", "type"],
+     *          rule: { type: "boolean" },
+     *          message: "validation-invalid-field-type"
+     *      }
+     *  ]
+     * }
+     *
+     * Note that as in the examples above, the GSS metaschema defines "error hints" that result in message keys
+     * being returned.
+     *
+     * NOTE: This function is a non-API function, i.e. one that assists public functions in performing their work, but
+     * which is not guaranteed to remain available.
+     *
+     * @param {Object} gssSchema - A GSS schema.
+     * @param {ajvErrors|Boolean} ajvErrors - The raw errors returned by AJV, if there are any, or `false` if there are no validation Errors.
+     * @return {Object} - An object detailing the validation results (see above).
```

Comment:
@typedef for this validation return payload here using the newly created @ajvError type

Reply:
Done.